x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStore.java

```diff
@@ -83,47 +100,120 @@
     private final Settings settings;
     private final Client client;
     private final SecurityIndexManager securityIndexManager;
+    private final Cache<String, Set<ApplicationPrivilegeDescriptor>> descriptorsCache;
+    private final CacheIteratorHelper<String, Set<ApplicationPrivilegeDescriptor>> descriptorsCacheHelper;
+    private final Cache<Set<String>, Set<String>> applicationNamesCache;
+    private final CacheIteratorHelper<Set<String>, Set<String>> applicationNamesCacheHelper;
+    private final AtomicLong numInvalidation = new AtomicLong();
 
     public NativePrivilegeStore(Settings settings, Client client, SecurityIndexManager securityIndexManager) {
         this.settings = settings;
         this.client = client;
         this.securityIndexManager = securityIndexManager;
+        CacheBuilder<String, Set<ApplicationPrivilegeDescriptor>> builder = CacheBuilder.builder();
+        final int cacheSize = DESCRIPTOR_CACHE_SIZE_SETTING.get(settings);
+        if (cacheSize >= 0) {
+            builder.setMaximumWeight(cacheSize);
+            builder.weigher((k, v) -> v.size());
+        }
+        descriptorsCache = builder.build();
+        descriptorsCacheHelper = new CacheIteratorHelper<>(descriptorsCache);
+
+        CacheBuilder<Set<String>, Set<String>> applicationNamesCacheBuilder = CacheBuilder.builder();
+        final int nameCacheSize = APPLICATION_NAME_CACHE_SIZE_SETTING.get(settings);
+        if (nameCacheSize >= 0) {
+            applicationNamesCacheBuilder.setMaximumWeight(nameCacheSize);
+            applicationNamesCacheBuilder.weigher((k, v) -> k.size() + v.size());
+        }
+        applicationNamesCache = applicationNamesCacheBuilder.build();
+        applicationNamesCacheHelper = new CacheIteratorHelper<>(applicationNamesCache);
     }
 
     public void getPrivileges(Collection<String> applications, Collection<String> names,
                               ActionListener<Collection<ApplicationPrivilegeDescriptor>> listener) {
+
+        // TODO: We should have a way to express true Zero applications
+        final Set<String> applicationNamesCacheKey = (isEmpty(applications) || applications.contains("*")) ?
+            Collections.singleton("*") : Set.copyOf(applications);
+
+        // Always fetch for the concrete application names even when the passed in application names do not
+        // contain any wildcard. This serves as a negative lookup.
+        Set<String> concreteApplicationNames = applicationNamesCache.get(applicationNamesCacheKey);
+
+        if (concreteApplicationNames != null && concreteApplicationNames.size() == 0) {
+            logger.debug("returning empty application privileges as application names result in empty list");
+            listener.onResponse(Collections.emptySet());
+
+        } else {
+            final Tuple<Set<String>, Map<String, Set<ApplicationPrivilegeDescriptor>>> cacheStatus;
```

Comment:
This complexity is due to an attempt to optimize number of documents to be fetched from index. It can be simplified if we always fetching everything when hitting the index is unavoidable (as you suggested below).

Reply:
Once we remove this optimization the `Tuple<Set<String>, Map<String, Set<ApplicationPrivilegeDescriptor>>>` data structure is no longer necessary. So complexity will definitely be reduced